File: 2023/day20/cpp/solution.cpp
// Day 20: Pulse Propagation - Module communication simulation
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <numeric>

enum class ModuleType { BROADCASTER, FLIP_FLOP, CONJUNCTION };

struct Module {
    ModuleType type;
    std::vector<std::string> destinations;
    bool state = false;  // For flip-flops
    std::unordered_map<std::string, bool> memory;  // For conjunctions
};

struct Pulse {
    std::string source;
    std::string dest;
    bool high;
};

std::unordered_map<std::string, Module> parse_input(const std::string& filename) {
    std::unordered_map<std::string, Module> modules;
    std::ifstream file(filename);
    std::string line;

    while (std::getline(file, line)) {
        if (line.empty()) continue;

        size_t arrow_pos = line.find(" -> ");
        std::string name_part = line.substr(0, arrow_pos);
        std::string dest_part = line.substr(arrow_pos + 4);

        Module module;
        std::string name;

        // Parse destinations
        std::stringstream ss(dest_part);
        std::string dest;
        while (std::getline(ss, dest, ',')) {
            // Trim whitespace
            size_t start = dest.find_first_not_of(" ");
            size_t end = dest.find_last_not_of(" ");
            if (start != std::string::npos) {
                module.destinations.push_back(dest.substr(start, end - start + 1));
            }
        }

        if (name_part == "broadcaster") {
            name = "broadcaster";
            module.type = ModuleType::BROADCASTER;
        } else if (name_part[0] == '%') {
            name = name_part.substr(1);
            module.type = ModuleType::FLIP_FLOP;
            module.state = false;
        } else if (name_part[0] == '&') {
            name = name_part.substr(1);
            module.type = ModuleType::CONJUNCTION;
        }

        modules[name] = module;
    }

    // Initialize conjunction memory for all inputs
    for (auto& [name, module] : modules) {
        for (const auto& dest : module.destinations) {
            if (modules.count(dest) && modules[dest].type == ModuleType::CONJUNCTION) {
                modules[dest].memory[name] = false;
            }
        }
    }

    return modules;
}

void reset_state(std::unordered_map<std::string, Module>& modules) {
    for (auto& [name, module] : modules) {
        if (module.type == ModuleType::FLIP_FLOP) {
            module.state = false;
        } else if (module.type == ModuleType::CONJUNCTION) {
            for (auto& [key, val] : module.memory) {
                val = false;
            }
        }
    }
}

struct SimResult {
    long long low_count;
    long long high_count;
    std::unordered_set<std::string> high_senders;
};

SimResult simulate_button_press(std::unordered_map<std::string, Module>& modules,
                                 const std::unordered_set<std::string>& watch_nodes = {}) {
    long long low_count = 0;
    long long high_count = 0;
    std::unordered_set<std::string> high_senders;

    std::queue<Pulse> q;
    q.push({"button", "broadcaster", false});

    while (!q.empty()) {
        Pulse pulse = q.front();
        q.pop();

        if (pulse.high) {
            high_count++;
        } else {
            low_count++;
        }

        // Track if watched nodes send high pulses
        if (!watch_nodes.empty() && watch_nodes.count(pulse.source) && pulse.high) {
            high_senders.insert(pulse.source);
        }

        if (!modules.count(pulse.dest)) {
            continue;
        }

        Module& module = modules[pulse.dest];

        if (module.type == ModuleType::BROADCASTER) {
            for (const auto& next_dest : module.destinations) {
                q.push({pulse.dest, next_dest, pulse.high});
            }
        } else if (module.type == ModuleType::FLIP_FLOP) {
            if (!pulse.high) {  // Only react to low pulses
                module.state = !module.state;
                for (const auto& next_dest : module.destinations) {
                    q.push({pulse.dest, next_dest, module.state});
                }
            }
        } else if (module.type == ModuleType::CONJUNCTION) {
            module.memory[pulse.source] = pulse.high;
            // Send low if all inputs are high, otherwise send high
            bool all_high = true;
            for (const auto& [key, val] : module.memory) {
                if (!val) {
                    all_high = false;
                    break;
                }
            }
            bool output = !all_high;
            for (const auto& next_dest : module.destinations) {
                q.push({pulse.dest, next_dest, output});
            }
        }
    }

    return {low_count, high_count, high_senders};
}

long long part1(std::unordered_map<std::string, Module> modules) {
    reset_state(modules);

    long long total_low = 0;
    long long total_high = 0;

    for (int i = 0; i < 1000; i++) {
        auto result = simulate_button_press(modules);
        total_low += result.low_count;
        total_high += result.high_count;
    }

    return total_low * total_high;
}

long long part2(std::unordered_map<std::string, Module> modules) {
    reset_state(modules);

    // Find the module that feeds into rx
    std::string rx_input;
    for (const auto& [name, module] : modules) {
        for (const auto& dest : module.destinations) {
            if (dest == "rx") {
                rx_input = name;
                break;
            }
        }
        if (!rx_input.empty()) break;
    }

    if (rx_input.empty()) {
        return 0;
    }

    // Find all modules that feed into rx_input
    std::unordered_set<std::string> watch_nodes;
    for (const auto& [key, val] : modules[rx_input].memory) {
        watch_nodes.insert(key);
    }

    std::unordered_map<std::string, long long> cycle_lengths;

    long long button_press = 0;
    while (cycle_lengths.size() < watch_nodes.size()) {
        button_press++;
        auto result = simulate_button_press(modules, watch_nodes);

        for (const auto& node : result.high_senders) {
            if (!cycle_lengths.count(node)) {
                cycle_lengths[node] = button_press;
            }
        }
    }

    // LCM of all cycle lengths
    long long result = 1;
    for (const auto& [node, length] : cycle_lengths) {
        result = std::lcm(result, length);
    }

    return result;
}

int main() {
    auto modules = parse_input("../input.txt");

    std::cout << "Part 1: " << part1(modules) << std::endl;
    std::cout << "Part 2: " << part2(modules) << std::endl;

    return 0;
}

File: 2023/day20/go/solution.go
// Day 20: Pulse Propagation - Module communication simulation
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

type ModuleType int

const (
	Broadcaster ModuleType = iota
	FlipFlop
	Conjunction
)

type Module struct {
	moduleType   ModuleType
	destinations []string
	state        bool            // For flip-flops: on/off
	memory       map[string]bool // For conjunctions: last pulse from each input
}

type Pulse struct {
	source string
	dest   string
	high   bool
}

func parseInput(filename string) map[string]*Module {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	modules := make(map[string]*Module)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, " -> ")
		namePart := parts[0]
		destPart := parts[1]

		destinations := make([]string, 0)
		for _, d := range strings.Split(destPart, ", ") {
			destinations = append(destinations, strings.TrimSpace(d))
		}

		if namePart == "broadcaster" {
			modules["broadcaster"] = &Module{
				moduleType:   Broadcaster,
				destinations: destinations,
			}
		} else if strings.HasPrefix(namePart, "%") {
			name := namePart[1:]
			modules[name] = &Module{
				moduleType:   FlipFlop,
				destinations: destinations,
				state:        false,
			}
		} else if strings.HasPrefix(namePart, "&") {
			name := namePart[1:]
			modules[name] = &Module{
				moduleType:   Conjunction,
				destinations: destinations,
				memory:       make(map[string]bool),
			}
		}
	}

	// Initialize conjunction memory for all inputs
	for name, module := range modules {
		for _, dest := range module.destinations {
			if destModule, exists := modules[dest]; exists && destModule.moduleType == Conjunction {
				destModule.memory[name] = false // false = low pulse
			}
		}
	}

	return modules
}

func resetState(modules map[string]*Module) {
	for _, module := range modules {
		if module.moduleType == FlipFlop {
			module.state = false
		} else if module.moduleType == Conjunction {
			for key := range module.memory {
				module.memory[key] = false
			}
		}
	}
}

func simulateButtonPress(modules map[string]*Module, watchNodes map[string]bool) (int, int, map[string]bool) {
	lowCount := 0
	highCount := 0
	highSenders := make(map[string]bool)

	// Queue for BFS
	queue := []Pulse{{source: "button", dest: "broadcaster", high: false}}

	for len(queue) > 0 {
		pulse := queue[0]
		queue = queue[1:]

		if pulse.high {
			highCount++
		} else {
			lowCount++
		}

		// Track if watched nodes send high pulses
		if watchNodes != nil && watchNodes[pulse.source] && pulse.high {
			highSenders[pulse.source] = true
		}

		module, exists := modules[pulse.dest]
		if !exists {
			continue
		}

		switch module.moduleType {
		case Broadcaster:
			for _, nextDest := range module.destinations {
				queue = append(queue, Pulse{source: pulse.dest, dest: nextDest, high: pulse.high})
			}

		case FlipFlop:
			if !pulse.high { // Only react to low pulses
				module.state = !module.state
				for _, nextDest := range module.destinations {
					queue = append(queue, Pulse{source: pulse.dest, dest: nextDest, high: module.state})
				}
			}

		case Conjunction:
			module.memory[pulse.source] = pulse.high
			// Send low if all inputs are high, otherwise send high
			output := false
			for _, val := range module.memory {
				if !val {
					output = true
					break
				}
			}
			for _, nextDest := range module.destinations {
				queue = append(queue, Pulse{source: pulse.dest, dest: nextDest, high: output})
			}
		}
	}

	return lowCount, highCount, highSenders
}

func part1(modules map[string]*Module) int {
	resetState(modules)

	totalLow := 0
	totalHigh := 0

	for i := 0; i < 1000; i++ {
		low, high, _ := simulateButtonPress(modules, nil)
		totalLow += low
		totalHigh += high
	}

	return totalLow * totalHigh
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func lcm(a, b int) int {
	return a / gcd(a, b) * b
}

func part2(modules map[string]*Module) int {
	resetState(modules)

	// Find the module that feeds into rx
	var rxInput string
	for name, module := range modules {
		for _, dest := range module.destinations {
			if dest == "rx" {
				rxInput = name
				break
			}
		}
		if rxInput != "" {
			break
		}
	}

	if rxInput == "" {
		return 0
	}

	// Find all modules that feed into rxInput
	watchNodes := make(map[string]bool)
	for name := range modules[rxInput].memory {
		watchNodes[name] = true
	}

	cycleLengths := make(map[string]int)

	buttonPress := 0
	for len(cycleLengths) < len(watchNodes) {
		buttonPress++
		_, _, highSenders := simulateButtonPress(modules, watchNodes)

		for node := range highSenders {
			if _, found := cycleLengths[node]; !found {
				cycleLengths[node] = buttonPress
			}
		}
	}

	// LCM of all cycle lengths
	result := 1
	for _, length := range cycleLengths {
		result = lcm(result, length)
	}

	return result
}

func main() {
	_, currentFile, _, _ := runtime.Caller(0)
	inputPath := filepath.Join(filepath.Dir(currentFile), "..", "input.txt")

	modules := parseInput(inputPath)
	fmt.Printf("Part 1: %d\n", part1(modules))

	// Re-parse for part 2 (fresh state)
	modules = parseInput(inputPath)
	fmt.Printf("Part 2: %d\n", part2(modules))
}
